### ISSUE TITLE
Routing Regular Matching Issue in Kong 3.0.x

### ISSUE THREAD

**Issue by xiupengrong** [Kong/kong#12082]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

3.5.0

### Current Behavior

The same routing regular expression in version 2.8.0 as in version 3.5.0 does not work in version 3.5.0
The routing regular expression is : /api/[0-9a-zA-Z]*/((v0)|(v1)|(v2))(/[\s\S]*)*

**Comment by xiupengrong** [comment]
the mode of the router is traditional_compat

**Comment by Water-Melon** [comment]
Could you provide a simple example to reproduce the issue? And could you offer more detailed information about the failure?

**Comment by xiupengrong** [comment]
> I think there may have been some changes. Please refer to versions [2.8](https://docs.konghq.com/gateway/latest/key-concepts/routes/#regular-expressions) and [3.5](https://docs.konghq.com/gateway/latest/key-concepts/routes/expressions/) for comparison.我认为可能发生了一些变化。请参考2.8和3.5版本进行比较。

I'd like to know what has changed in terms of router regular expressions and what is not compatible, I don't seem to have seen a description of it, I think traditional_compat mode seems to be compatible with version 2.8.0

**Comment by chronolaw** [comment]
No, both `traditional_compatible ` and `expressions` are base on `atc-router`, which is built by Rust, that means the expression use Rust regex library, not Nginx PCRE.

**Comment by xiupengrong** [comment]
> No, both `traditional_compatible ` and `expressions` are base on `atc-router`, which is built by Rust, that means the expression use Rust regex library, not Nginx PCRE.不， `traditional_compatible ` 和 `expressions` 都是基于 Rust 构建的 `atc-router` ，这意味着表达式使用 Rust 正则表达式库，而不是 Nginx PCRE。

So that means totally incompatible, where can rust regular expression library be learned, can you give me a link, thanks.

**Comment by chronolaw** [comment]
I think it is 99% compatible with PCRE, the doc is here: https://docs.rs/regex/latest/regex/index.html

**Comment by bungle** [comment]
@xiupengrong, you can always use `traditional` router too, which is still shipped and it is the same as before.

**Comment by xiupengrong** [comment]
> @xiupengrong, you can always use `traditional` router too, which is still shipped and it is the same as before (just not the default one anymore).，您也可以随时使用 `traditional` 路由器，该路由器仍在发货，并且与以前相同（只是不再是默认路由器）。

bungle, thanks for your reply, I set router_flavor = "traditional", it's still not the same as expected from 2.8.0, compared to "/api" and "/api/[0-9a-zA-Z]/((v0)|(v1)|(v1)(/[\s\S])*", it will match "/api" preferentially if modified to "~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*", it will match "~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*", what is the meaning of "~"? My routing in version 2.8.0 is not configured with "~"

**Comment by xiupengrong** [comment]
> @xiupengrong, you can always use `traditional` router too, which is still shipped and it is the same as before (just not the default one anymore).

bungle, thanks for your reply, I set router_flavor = "traditional", it's still not the same as expected from 2.8.0, compared to "/api" and "/api/[0-9a-zA-Z]/((v0)|(v1)|(v1)(/[\s\S])*", it will match "/api" preferentially if modified to ‘’~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*”, it will match‘’~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*”, what is the meaning of "~"? My routing in version 2.8.0 is not configured with "~"

**Comment by bungle** [comment]
> > @xiupengrong, you can always use `traditional` router too, which is still shipped and it is the same as before (just not the default one anymore).
> 
> bungle, thanks for your reply, I set `router_flavor = "traditional"`, it's still not the same as expected from 2.8.0, compared to `/api` and `/api/[0-9a-zA-Z]/((v0)|(v1)|(v1)(/[\s\S])*`, it will match `/api `preferentially , if modified to `~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*`, it will match`~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*`，what is the meaning of `~`? My routing in version 2.8.0 is not configured with "~"

Yes, we changed the format so that regexes need to have `~`-prefix. You also need to use that when creating new routes. Otherwise they will be treated as non-regex prefixes.

**Comment by xiupengrong** [comment]
> > > @xiupengrong, you can always use `traditional` router too, which is still shipped and it is the same as before (just not the default one anymore).
> > 
> > 
> > bungle, thanks for your reply, I set `router_flavor = "traditional"`, it's still not the same as expected from 2.8.0, compared to `/api` and `/api/[0-9a-zA-Z]/((v0)|(v1)|(v1)(/[\s\S])*`, it will match `/api `preferentially , if modified to `~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*`, it will match`~/api/[0-9a-zA-Z]/((v0)|(v1)|(v2))(/[\s\S])*`，what is the meaning of `~`? My routing in version 2.8.0 is not configured with "~"
> 
> Yes, we changed the format so that regexes need to have `~`-prefix. Existing routes are automatically migrated. You also need to use that when creating new routes. Otherwise they will be treated as non-regex prefixes. The original regex detection was too magical, thus this change.

All right, thanks.

**Comment by bungle** [comment]
Closing this, as it looks to be resolved.
